### DISCUSSION TITLE
Making Registries Easier to Use

### DISCUSSION THREAD

**Discussion by reitowo** [microsoft/vcpkg#17132] [Ideas]
We have to
1. Commit our source repo
2. Get commit id
3. Update it in 'ports/(portname)/portfile.cmake'
4. Use overlay feature and '--head' to get SHA512, or need to update git-tree in versions, which is more annoying
5. Update SHA512
6. Use 'rev-parse HEAD:ports/xx' to get git-tree object id
7. Update in versions/x-/xx.json and baseline.json

😅 Too complicated to do this every time

Update: An python script to do the dirty work. Just put it under registry's root folder (Please check https://github.com/cnSchwarzer/vcpkg-ports) and run it use python.exe ./auto-update.py. A flag -f will force update all ports. This only works with vcpkg_from_github.

Code: https://gist.github.com/cnSchwarzer/7922a542eb192fdf6623e48e77664bec

> **Don't forget to remove or change the proxy setting in the code**

**Comment by strega-nil** [comment]
So, I agree in theory; however, I'm not sure exactly what to do about it, and I think this needs design work. I think more information is needed on how you might like this process to work, as well as what others might like: what might make a difference in your case, what features we could create, etc. I think this is better suited for a discussion, where people can make suggestions and we can spin out PRs and issues from that discussion, so I'll move it over ~~there~~ here :)

**Comment by reitowo** [comment]
Maybe a simple automation script will work. 
1. We still manually push our source repos (In my case `https://github.com/cnSchwarzer/bup`)
2. Put a script in private `vcpkg-ports` repo's root (In my case `https://github.com/cnSchwarzer/vcpkg-ports`), which automatically does following jobs.
3. Scan all private repos in `ports` folder (In my case `bup`, `k1ee`, etc.) or specify ports for following operations.
4. Get remote repos' latest commit id and compare with the ref-id in `vcpkg_from_github` command
5. If they differ, update the ref-id and download the source zip, automatically calculate SHA512.
6. Increase minor/subminor/buildnumber by one in `vcpkg.json`
7. Commit and push `ports` folder
8. `rev-parse HEAD:ports/xxx` updated ports to get updated git-tree object id.
9. Update them in `versions/x-/xx.json`
10. Update new version number in `version/baseline.json`

This only considered `vcpkg_from_github` method.
I know it seems every user could build their own script, but still vcpkg official can provide such scripts in helper tools.

**Reply by autoantwort** [reply]
afaik you can do step 8,9,10 by running `./vcpkg x-add-version --all`

**Reply by reitowo** [reply]
@autoantwort But how to update in a private port folder? This command seems only handle ports in root folder.

**Reply by autoantwort** [reply]
You can specify `--x-builtin-ports-root=...` (normally `ports`) and `--x-builtin-registry-versions-dir=...` (normally `versions`), but to be honest my expectation was that you don't have to specify it when you are in the root of https://github.com/cnSchwarzer/vcpkg-ports. Maybe you have to add `.vcpkg-root` file there, idk

**Reply by reitowo** [reply]
Yes, I also expect that. Anyway thanks!

**Reply by strega-nil** [reply]
hmm, I think that this is just an issue with `x-add-version`; in the original specification, `x-add-version` would work in the registry you're cd'd into, so we could extend it that way.

**Comment by reitowo** [comment]
Check https://github.com/microsoft/vcpkg-tool/pull/52

**Comment by reitowo** [comment]
An python script to do the dirty work. Just put it under registry's root folder (Please check https://github.com/cnSchwarzer/vcpkg-ports) and run it use `python.exe ./auto-update.py`. A flag `-f` will force update all ports. This only works with `vcpkg_from_github`.

Code: https://gist.github.com/cnSchwarzer/7922a542eb192fdf6623e48e77664bec
> **Don't forget to remove or change the proxy setting in the code**

**Comment by DimRochette** [comment]
still unable to `x-add-version` on custom registry we use an ugly powershell script here to do something worse

**Reply by DimRochette** [reply]
it works for me that way, from root of private registry
```
vcpkg.exe format-manifest ports/yourport/vcpkg.json
vcpkg.exe x-add-version --all --overlay-ports=./ports --x-builtin-registry-versions-dir=./versions/ --x-builtin-ports-root=./ports
```
